### PR TITLE
Record: 11L LeakyReLU² XSA-all GPTQ-AR SLOT64 — 0.6951 BPB

### PR DESCRIPTION
**Pull request by canivel**
## Summary

- **SLOT BPB: 0.6951** (seed 1337, 8×H100 SXM)
- Sliding window BPB: 1.1316
- Pre-quant BPB: 1.0508
- Artifact: 15.69 MB (under 16 MB limit)
- Training: 5,684 steps in 595s (104.69 ms/step)

## Key Innovations

1. **SLOT with Warmstart (α=0.85)**: Per-window learned delta + logit bias optimized via 64 AdamW steps. Adjacent windows inherit 85% of optimizer state, dramatically accelerating convergence.

2. **AR Self-Generated GPTQ Calibration**: Model generates its own 64×2048 calibration tokens autoregressively (temp=0.8) — no training data leakage, better distribution match for quantization.

3. **Full Hessian GPTQ**: Cholesky error compensation + column reordering by Hessian diagonal. Int6 MLP+attn, Int8 embed, lzma preset=9.

## Architecture
11L 512d GQA(8h/4kv), LeakyReLU(0.5)² MLP 3x, XSA on all 11 layers, BigramHash(3072,112), focal loss γ=1.0, sqrt warmdown, Parallel Muon + EMA + SWA.

## Known Limitations

- **Single seed run** — budget constraints. Seed 42 training completed (pre-quant 1.0525 BPB) but pod terminated during GPTQ.
- **SLOT eval takes 825s** (13.7 min) — exceeds the 10-min eval budget. Reducing to SLOT_STEPS=46 (~593s) is straightforward and planned for follow-up with 3-seed validation.
- Requesting additional compute grant for 3-seed validation.

## Test Plan
- [ ] Verify `train_gpt.py` compiles and runs on 8×H100
- [ ] Confirm artifact ≤ 16 MB
- [ ] Validate SLOT BPB with reduced steps (46) fits in 10-min eval
- [ ] 3-seed validation with p < 0.01

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — 11L LeakyReLU² XSA-all GPTQ-AR SLOT64

**BPB:** 0.6951 (SLOT) | 1.1316 (sliding-window, no SLOT) | **Seeds:** 1 (seed 1337) | **Artifact:** 15.69 MB | **Compliance:** COMPLIANCE FLAG

**What this does:** 11L / 512d / GQA(8h/4kv) with LeakyReLU(0.5)² MLP, XSA on all layers, Full-Hessian GPTQ (AR self-generated calibration), and — for the headline number — a 64-step per-window "SLOT" test-time optimization that learns a per-sample hidden-state `delta` and an output `logit_bias` at eval time, with an α=0.85 warmstart of optimizer state across adjacent windows.

**What I found in the code** (`records/track_10min_16mb/2026-04-03_11L_LeakyReLU2_XSA-all_GPTQ-AR_SLOT64_0.6951/train_gpt.py` @ SHA `7296c4da1fe6a6bf4d23a41afcf1ecf83a589453`):

The SLOT inner loop is `eval_val_sliding_slot` at lines 846-945. The critical region is lines 897-921:

```python
# lines 897-901 — mask = scored positions only
mask = torch.zeros(bsz, seq_len, device=device)
for i, ws in enumerate(batch_ws):
    wlen = wlens[i]
    s = 0 if ws == 0 else max(wlen - stride, 0)
    mask[i, s:wlen] = 1.0            # mask covers [wlen-stride, wlen): the scored region
valid_count = mask.sum()
...
# lines 912-921 — inner optimization
for _step in range(slot_steps):       # slot_steps = 64
    ...
    h = hidden_f + delta
    logits_proj = F.linear(h, proj_w) + logit_bias
    logits = softcap * torch.tanh(logits_proj / softcap)
    nll_opt = F.cross_entropy(logits.reshape(-1, logits.size(-1)),
                              targets_flat, reduction="none").reshape(bsz, seq_len)
    slot_loss = (nll_opt * mask).sum() / valid_count   # loss is over scored positions
    slot_opt.zero_grad(); slot_loss.backward(); slot_opt.step()
# lines 926-935 — score happens AFTER the 64 inner steps, on the same [s:wlen] range
```

- **Position scope of the inner loss:** the mask is exactly `[wlen-stride, wlen)` — i.e. the scored positions. The SLOT loop is **not** restricted to strictly-prior context; it optimizes `delta` and `logit_bias` directly against the cross-entropy at the same positions that are then scored. This is **standard SLOT**, not causal/context-only SLOT.
- **Score-before-update:** violated. Lines 912-921 run 64 AdamW updates on `delta`/`logit_bias` against the target cross-entropy, and only after all 64 steps does line 930 compute the scored `nll` (still using the updated parameters).
- **Single L→R pass:** violated. 64 inner passes per window over the same tokens; scoring uses only the final pass.
- **Warmstart (α=0.85, lines 904-906) + stride=64 / seq_len=2048:** adjacent windows share 1984 tokens, and the optimizer state (scaled delta + logit_bias) carries forward, so a given target token's prediction is influenced by optimization on that same token in a previous overlapping window — further entangling update and score.
- **Effect on reported BPB:** sliding-window (no SLOT) is 1.1316; add SLOT and it drops to 0.6951. A ~0.44 BPB drop from 64 inner AdamW steps against the scored positions is the expected signature of fitting the target distribution at eval, not a legal inference-time improvement.
- **Per Issue #1017 4-conditions:** (1) causal dependence — FAIL; (2) full normalized distribution — OK; (3) score-before-update — FAIL; (4) single L→R pass — FAIL.

**CPU gauntlet** (CPU-only): PASS. Import/Hyperparameters/Model/Forward all PASS. Artifact 4,638,699 bytes (29.0% of 16MB) after int6+lzma. Est. 8×H100 ~13,022 steps in 10 min. So the code itself is well-formed and fits budget — the flag is purely about SLOT legality, not artifact or training.

**Other concerns (secondary):**
- **Single-seed**. PR body + README both disclose only seed 1337; the 3-seed requirement is unmet.
- **Eval budget**. Author self-reports SLOT eval at 825s on 8×H100, exceeding the 600s eval budget ("SLOT eval takes 825s (13.7 min) — exceeds the 10-min eval budget", PR body). Any reproduction of 0.6951 would not complete in the eval image.

**Questions/flags:**
- Per Issue #1017 conditions: (1) causal dependence, (2) full normalized distribution, (3) score-before-update, (4) single L→R pass. As traced above, (1), (3), and (4) are violated by `eval_val_sliding_slot`.
- SLOT legality is pending per Issue #1336. **Standard SLOT (optimizing all positions) was flagged as illegal; causal/context-only SLOT awaits ruling.** The code here is standard SLOT — the inner loss is taken on the scored positions (`mask = [wlen-stride, wlen)`), not on strictly-prior context. Even if Issue #1336 lands permissive for *causal* SLOT, this submission would still be outside that scope.
- Per Issue #677 (valerio-oai), multi-epoch TTT/SLOT that scores only on the final pass is invalid. The 64-step inner loop with scoring after the last update is the same pattern.
- The legal sliding-window number from the same model is **1.1316 BPB**, which is above the current merged SOTA (~1.0810) — this submission does not beat SOTA without the SLOT step.

**Verdict:** COMPLIANCE FLAG — standard SLOT (not causal/context-only). The 0.6951 number is produced by 64 AdamW steps that optimize `delta`/`logit_bias` against the same tokens that are subsequently scored, violating Issue #1017 conditions (1), (3), and (4). Per Issue #677 the multi-epoch-then-score pattern is also independently invalid.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:**
- **CLOSE or HOLD pending Issue #1336** — but note that even a permissive ruling for *causal* SLOT would not cover this PR, because the inner loss in `eval_val_sliding_slot` lines 897-921 is computed on the scored positions, not on strictly-prior context. The pre-SLOT sliding-window BPB of 1.1316 is the legally-comparable number for this submission and is above current SOTA.
- Separately, the author discloses eval wallclock 825s > 600s budget and only 1 seed, so even if the SLOT method were ruled legal, the submission as-is would still need a ≤600s eval config and 3-seed validation.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU gauntlet PASS: artifact 4.64 MB (29.0% of 16 MB budget), forward pass OK, est. 13,022 steps on 8×H100 in 10 min. AI tooling: review drafted with Claude Code (Sonnet/Opus) using an internal review template; all citations, file paths, and compliance audits were verified against the PR's actual code at SHA `7296c4da1fe6a6bf4d23a41afcf1ecf83a589453`.*

**Comment by Bortlesboat** [comment]
Automated compliance check flagged a SLOT-style adapt-then-score pattern. This matches the **SLOT portion** of @MatoTeziTanka's ruling on #1376 ("FLAG (two independent violations)" — SLOT + Pre-quant TTT). Posting line-level evidence.

**Rule (from issue #1017):** "For any token in val, the model state used to predict it must be determined only by data seen strictly before it." Optimizing any scoring parameter (delta, logit bias, per-sample embedding, etc.) against the val targets that are about to be scored is equivalent to training on val — the optimizer is *choosing* the parameters that minimize loss on those exact tokens.

**Evidence — `eval_val_sliding_slot` (lines 846–940):**

```python
# line 846
def eval_val_sliding_slot(args, base_model, rank, world_size, device,
                          val_tokens, ...) -> tuple[float, float]:
    """Sliding window eval with SLOT: per-sample delta + logit bias optimization."""
    ...
    for bi in range(0, len(my_windows), batch_seqs):
        ...
        # lines 886-893 — build x_batch / y_batch from val_tokens
        for i, ws in enumerate(batch_ws):
            end = min(ws + seq_len, total_tokens)
            wlen = end - ws
            ...
            chunk = val_tokens[ws:end + 1].to(dtype=torch.int64, device=device)
            x_batch[i, :wlen] = chunk[:-1]
            y_batch[i, :wlen] = chunk[1:]

        # lines 894-895 — hidden states from the frozen base model
        with torch.no_grad(), torch.autocast(device_type="cuda", dtype=torch.bfloat16):
            hidden = compiled_hidden(x_batch)
        hidden_f = hidden.float()
        ...

        # lines 908-910 — create per-sample delta + logit bias, both requires_grad=True
        delta = torch.zeros(bsz, 1, hidden_f.size(-1), device=device,
                            dtype=torch.float32, requires_grad=True)
        logit_bias = torch.zeros(bsz, 1, proj_w.size(0), device=device,
                                 dtype=torch.float32, requires_grad=True)
        slot_opt = torch.optim.AdamW([delta, logit_bias], lr=slot_lr)
        targets_flat = y_batch.reshape(-1)   # the val targets we're about to score

        # lines 912-921 — SLOT_STEPS=64 AdamW steps minimizing loss AGAINST y_batch
        for _step in range(slot_steps):
            ...
            h = hidden_f + delta
            logits_proj = F.linear(h, proj_w) + logit_bias
            logits = softcap * torch.tanh(logits_proj / softcap)
            nll_opt = F.cross_entropy(logits.reshape(-1, logits.size(-1)),
                                      targets_flat, reduction="none").reshape(bsz, seq_len)
            slot_loss = (nll_opt * mask).sum() / valid_count
            slot_opt.zero_grad()
            slot_loss.backward()
            slot_opt.step()

        # lines 926-931 — recompute nll using the delta/logit_bias that were just
        # optimized against those exact targets, and report it as the score
        with torch.no_grad():
            h = hidden_f + delta
            logits_proj = F.linear(h, proj_w) + logit_bias
            logits = softcap * torch.tanh(logits_proj / softcap)
            nll = F.cross_entropy(logits.reshape(-1, logits.size(-1)),
                                  targets_flat, reduction="none").reshape(bsz, seq_len)
        ...
        # line 936 — accumulate into loss_sum
        loss_sum += scored_nll.sum()
```

Defaults confirm this is on by default (line 90: `SLOT_STEPS=64`, line 89: `SLOT_ENABLED=1`).

**Why this is a C3 violation:**

For each batch of val tokens, `delta` and `logit_bias` are learnable parameters that are optimized via AdamW to minimize cross-entropy *on the very targets* (`y_batch`) that are about to be scored. The "final" nll at line 930 uses the optimizer's converged output — which by construction is at least as low as any no-train baseline, because the optimizer just performed 64 gradient steps to make it so. Every scored token was a supervised target for the parameters that score it.

The warm-start (lines 904–906) carries optimized `delta`/`logit_bias` from batch to batch, but that doesn't rescue the pattern — each batch still performs its own 64 AdamW steps against its own `y_batch` targets.

**Precedent:** This is exactly the same SLOT-style pattern flagged in #1376 by @MatoTeziTanka. The Pre-quant-TTT family (#1376, #1488, #1487) has already been closed as invalid. #1488 author @ndokutovich [confirmed and closed](https://github.com/openai/parameter-golf/pull/1488#issuecomment-4248289728) referencing "the #1376 cluster ruling."

**Source:** generated by [parameter-golf-checker](https://github.com/Bortlesboat/parameter-golf-checker) (context in #1603). Happy to correct if I'm misreading the control flow — `@canivel` please let me know if `SLOT_ENABLED=0` was set for the run that produced the reported BPB.
